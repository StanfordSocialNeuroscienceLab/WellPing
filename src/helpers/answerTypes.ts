import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
