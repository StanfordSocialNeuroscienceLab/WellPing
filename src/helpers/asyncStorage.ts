import { isToday, addDays } from "date-fns";
import { Notifications } from "expo";
import { AsyncStorage } from "react-native";

import { SurveyScreenState } from "../SurveyScreen";
import PingEntity from "../entities/PingEntity";
import {
  isTimeThisWeekAsync,
  getAllStreamNamesAsync,
  getSurveyFileAsync,
} from "./configFiles";
import { logError } from "./debug";
import { StreamName } from "./types";

export async function getASKeyAsync(key: string = ""): Promise<string> {
  const studyId = (await getSurveyFileAsync()).studyInfo.id;
  return `@WELLPING:Study_${studyId}/${key}`; //TODO: ADD ENCODED URL HERE
}

const NOTIFICATION_TIMES_KEY = `NotificationTime`;

export async function storeNotificationTimesAsync(times: Date[]) {
  try {
    await AsyncStorage.setItem(
      await getASKeyAsync(NOTIFICATION_TIMES_KEY),
      JSON.stringify(times),
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function clearNotificationTimesAsync() {
  try {
    await Notifications.cancelAllScheduledNotificationsAsync();
    await AsyncStorage.removeItem(await getASKeyAsync(NOTIFICATION_TIMES_KEY));
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function getNotificationTimesAsync(): Promise<Date[] | null> {
  try {
    const value = await AsyncStorage.getItem(
      await getASKeyAsync(NOTIFICATION_TIMES_KEY),
    );
    if (value == null) {
      return null;
    }
    const timesString: string[] = JSON.parse(value);
    const times = timesString.map((dateString) => new Date(dateString));
    return times;
  } catch (error) {
    // Error retrieving data
    logError(error);
    return null;
  }
}

/** TYPEPING TABLE (stored number of pings answered for each type) **/
export async function getNumberOfPingsForStreamName(
  streamName: StreamName,
): Promise<number> {
  const countRaw: {
    count: number;
  } = await PingEntity.createQueryBuilder()
    .select("COUNT(id) as count")
    .where("streamName = :streamName", { streamName })
    .getRawOne();
  return countRaw.count;
}

type NumbersOfPingsForAllStreamNames = {
  [stream: string /* actually StreamName */]: number;
};
export async function getNumbersOfPingsForAllStreamNames(): Promise<
  NumbersOfPingsForAllStreamNames
> {
  const pingsGroupByStreamNameRaw: {
    streamName: StreamName;
    count: number;
  }[] = await PingEntity.createQueryBuilder()
    .select("streamName, COUNT(id) as count")
    .groupBy("streamName")
    .getRawMany();

  return pingsGroupByStreamNameRaw.reduce((map, row) => {
    map[row.streamName] = row.count;
    return map;
  }, {} as NumbersOfPingsForAllStreamNames);
}

/** PING TABLE (stored started ping) **/
export async function insertPingAsync({
  notificationTime,
  startTime,
  streamName,
}: {
  notificationTime: Date;
  startTime: Date;
  streamName: StreamName;
}): Promise<PingEntity> {
  const newIndex = (await getNumberOfPingsForStreamName(streamName)) + 1;
  const pingId = `${streamName}${newIndex}`;
  const tzOffset = startTime.getTimezoneOffset();

  const pingEntity = new PingEntity();
  pingEntity.id = pingId;
  pingEntity.notificationTime = notificationTime;
  pingEntity.startTime = startTime;
  pingEntity.streamName = streamName;
  pingEntity.tzOffset = tzOffset;
  await pingEntity.save();

  return pingEntity;
}

export async function addEndTimeToPingAsync(
  pingId: string,
  endTime: Date,
): Promise<PingEntity> {
  const ping = await PingEntity.createQueryBuilder()
    .where("id = :pingId", { pingId })
    .getOne();
  if (ping == null) {
    throw new Error(`pingId ${pingId} not found in getPingsAsync.`);
  }

  ping.endTime = endTime;
  await ping.save();

  return ping;
}

export async function getLatestPingAsync(): Promise<PingEntity | null> {
  const latestPing = await PingEntity.createQueryBuilder()
    .orderBy("startTime", "DESC")
    .take(1)
    .getOne();
  if (!latestPing) {
    return null;
  }
  return latestPing;
}

export async function getPingsAsync(): Promise<PingEntity[]> {
  const pings = await PingEntity.createQueryBuilder()
    .orderBy("startTime", "DESC")
    .getMany();
  return pings;
}

export async function getTodayPingsAsync(): Promise<PingEntity[]> {
  const allPings = await getPingsAsync();
  const todayPings: PingEntity[] = [];
  // TODO: USE SQL HERE? (MAKE SURE TIMEZONE PROBLEM)
  for (const ping of allPings) {
    if (isToday(ping.notificationTime)) {
      todayPings.push(ping);
    }
    if (ping.notificationTime > addDays(new Date(), 1)) {
      break;
    }
  }
  return todayPings;
}

export async function getThisWeekPingsAsync(): Promise<PingEntity[]> {
  const allPings = await getPingsAsync();
  const thisWeekPings: PingEntity[] = [];
  // TODO: USE SQL HERE? (MAKE SURE TIMEZONE PROBLEM)
  for (const ping of allPings) {
    if (await isTimeThisWeekAsync(ping.notificationTime)) {
      thisWeekPings.push(ping);
    }
    if (ping.notificationTime > new Date()) {
      // Stop when the notification time is in the future.
      break;
    }
  }
  return thisWeekPings;
}

/** SURVEYSTATE TABLE (stores ping answers and other states) **/
const PINGS_STATE_PREFIX = `PingsState:`;

export async function storePingStateAsync(
  pingId: string,
  state: SurveyScreenState,
) {
  try {
    await AsyncStorage.setItem(
      `${await getASKeyAsync(PINGS_STATE_PREFIX)}${pingId}`,
      JSON.stringify(state),
    );
  } catch (error) {
    // Error saving data
    logError(error);
    throw error;
  }
}

export async function getPingStateAsync(
  pingId: string,
): Promise<SurveyScreenState> {
  try {
    const keyName = `${await getASKeyAsync(PINGS_STATE_PREFIX)}${pingId}`;
    const value = await AsyncStorage.getItem(keyName);
    if (value == null) {
      return {
        currentQuestionId: `ERROR: getPingStateAsync is null for ${keyName}`,
        extraMetaData: {},
        nextStack: [],
        currentQuestionAnswers: {},
        lastUploadDate: new Date(0),
      };
    }
    const state: SurveyScreenState = JSON.parse(value);
    if (state.lastUploadDate) {
      state.lastUploadDate = new Date(state.lastUploadDate);
    }
    return state;
  } catch (error) {
    // Error retrieving data
    logError(error);
    throw error;
  }
}

export async function clearPingStateAsync(pingId: string) {
  try {
    await AsyncStorage.removeItem(
      `${await getASKeyAsync(PINGS_STATE_PREFIX)}${pingId}`,
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

/** FUTURE PING QUEUE TABLE (used for storing follow-up streams) **/
const FUTURE_PING_QUEUE_KEY = `FuturePingQueue:`;

export type FuturePing = {
  afterDate: Date;
  streamName: StreamName;
};

type FuturePingJSON = {
  afterDate: string;
  streamName: string;
};

export async function initFuturePingQueueAsync() {
  try {
    await AsyncStorage.setItem(
      await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
      JSON.stringify([]),
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function enqueueToFuturePingQueue(futurePing: FuturePing) {
  const futurePings = await getFuturePingsQueue();
  if (futurePings == null) {
    throw new Error("futurePings == null in enqueueToFuturePingQueue");
  }
  futurePings.push(futurePing);
  //console.warn(`oy ${JSON.stringify(futurePings)}`);
  try {
    await AsyncStorage.setItem(
      await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
      JSON.stringify(futurePings),
    );
  } catch (error) {
    // Error saving data
    logError(error);
  }
}

export async function dequeueFuturePingIfAny(): Promise<FuturePing | null> {
  const futurePings = await getFuturePingsQueue();
  if (futurePings == null) {
    throw new Error("futurePings == null in dequeueFuturePingIfAny");
  }
  for (const futurePing of futurePings) {
    if (new Date() > futurePing.afterDate) {
      futurePings.shift();
      try {
        await AsyncStorage.setItem(
          await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
          JSON.stringify(futurePings),
        );
      } catch (error) {
        // Error saving data
        logError(error);
        return null;
      }
      return futurePing;
    }
  }
  return null;
}

export async function getFuturePingsQueue(): Promise<FuturePing[]> {
  try {
    const getValueAsync = async () => {
      return await AsyncStorage.getItem(
        await getASKeyAsync(FUTURE_PING_QUEUE_KEY),
      );
    };
    let value = await getValueAsync();
    if (value == null) {
      await initFuturePingQueueAsync();
      value = await getValueAsync();
    }

    if (value == null) {
      throw new Error(
        `${await getASKeyAsync(
          FUTURE_PING_QUEUE_KEY,
        )} is still null after initFuturePingQueueAsync()`,
      );
    }

    const futurePingsJSON: FuturePingJSON[] = JSON.parse(value);
    const futurePings: FuturePing[] = futurePingsJSON.map((futurePingJson) => ({
      ...futurePingJson,
      afterDate: new Date(futurePingJson.afterDate),
      streamName: futurePingJson.streamName,
    }));
    return futurePings;
  } catch (error) {
    // Error retrieving data
    logError(error);
    throw error;
  }
}
