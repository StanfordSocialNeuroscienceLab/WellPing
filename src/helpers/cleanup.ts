import { secureRemoveAllAnswersAsync } from "./secureStore/answer";
import { secureRemoveAllPingsAsync } from "./secureStore/ping";

export async function clearAllPingsAndAnswersAsync(): Promise<void> {
  await secureRemoveAllAnswersAsync();
  await secureRemoveAllPingsAsync();
}
