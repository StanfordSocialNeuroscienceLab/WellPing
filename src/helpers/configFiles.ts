import { isThisWeek } from "date-fns";

import { SurveyFile, Names, StudyID, StudyInfo, StreamName } from "./types";

export function getSurveyFile(): SurveyFile {
  const survey: SurveyFile = require("../../config/survey.json");
  return survey;
}

// TODO: DECOUPLE FUNCTIONS LIKE THIS
export function getNamesFile(): Names {
  const names: Names = require("../../config/names.json");
  return names;
}

export function getStudyInfo(): StudyInfo {
  return getSurveyFile().studyInfo;
}

export function getAllStreamNames(): StreamName[] {
  return Object.keys(getSurveyFile().meta.startingQuestionIds) as StreamName[];
}

export function getStudyId(): StudyID {
  return getStudyInfo().id;
}

export function isTimeThisWeek(time: Date): boolean {
  return isThisWeek(time, { weekStartsOn: getStudyInfo().weekStartsOn });
}
