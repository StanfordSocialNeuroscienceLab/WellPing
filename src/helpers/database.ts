import * as FileSystem from "expo-file-system";
import { Share } from "react-native";
import { createConnection, Connection, getConnection } from "typeorm";

import {
  AnswerEntity,
  SliderAnswerEntity,
  ChoicesWithSingleAnswerAnswerEntity,
  ChoicesWithMultipleAnswersAnswerEntity,
  YesNoAnswerEntity,
  MultipleTextAnswerEntity,
  HowLongAgoAnswerEntity,
} from "../entities/AnswerEntity";
import { PingEntity } from "../entities/PingEntity";
import { EncryptedColumnSubscriber } from "../helpers/typeormEncryption/EncryptedColumnSubscriber";
import {
  getCriticalProblemTextForUser,
  alertWithShareButtonContainingDebugInfo,
} from "./debug";

// TODO: remove me once typeorm is updated to a newer version (https://github.com/typeorm/typeorm/issues/6233)

const getDatabaseFilename = (databaseName: string) => `${databaseName}.db`;

export const entities = [
  AnswerEntity,
  SliderAnswerEntity,
  ChoicesWithSingleAnswerAnswerEntity,
  ChoicesWithMultipleAnswersAnswerEntity,
  YesNoAnswerEntity,
  MultipleTextAnswerEntity,
  HowLongAgoAnswerEntity,
  PingEntity,
];

export async function connectDatabaseAsync(
  databaseName: string,
): Promise<Connection> {
  try {
    return getConnection();
  } catch (e) {
    try {
      return await createConnection({
        type: "expo",
        driver: require("expo-sqlite"),
        database: getDatabaseFilename(databaseName),
        entities,
        subscribers: [EncryptedColumnSubscriber],
        synchronize: true,
        logging: __DEV__, // Only log in dev mode.
      });
    } catch (e) {
      alertWithShareButtonContainingDebugInfo(
        getCriticalProblemTextForUser(`connectDatabaseAsync: ${e}`),
      );
      throw e;
    }
  }
}

export function getDatabaseFolderUrl(): string {
  // https://forums.expo.io/t/how-to-upload-sqlite-db-dump/2315/3
  return `${FileSystem.documentDirectory}/SQLite/`;
}
export function getDatabaseFileUrl(databaseName: string): string {
  // https://forums.expo.io/t/how-to-upload-sqlite-db-dump/2315/3
  return `${getDatabaseFolderUrl()}${getDatabaseFilename(databaseName)}`;
}

export async function databaseFileExistsAsync(
  databaseName: string,
): Promise<boolean> {
  return (await FileSystem.getInfoAsync(getDatabaseFileUrl(databaseName)))
    .exists;
}

export async function shareDatabaseFileAsync(databaseName: string) {
  if (!(await databaseFileExistsAsync(databaseName))) {
    alert(`Database "${databaseName}" does not exists!`);
  }

  // https://forums.expo.io/t/how-to-upload-sqlite-db-dump/2315/3
  Share.share({
    url: getDatabaseFileUrl(databaseName),
  });
}

export async function deleteDatabaseFileAsync(databaseName: string) {
  if (!(await databaseFileExistsAsync(databaseName))) {
    return;
  }

  await FileSystem.deleteAsync(getDatabaseFileUrl(databaseName));
}

export async function backupDatabaseFileAsync(databaseName: string) {
  if (!(await databaseFileExistsAsync(databaseName))) {
    // Don't do anything if the database file does not exists.
    return;
  }

  await FileSystem.copyAsync({
    from: getDatabaseFileUrl(databaseName),
    to: getDatabaseFileUrl(`${databaseName}.${new Date().getTime()}.bk`),
  });
}

export async function getDatabaseFolderFilelistAsync() {
  return await FileSystem.readDirectoryAsync(getDatabaseFolderUrl());
}
