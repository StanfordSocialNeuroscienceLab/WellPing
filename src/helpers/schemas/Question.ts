import * as z from "zod";

import {
  StreamNameSchema,
  QuestionIdSchema,
  QuestionIdSchemaNullable,
} from "./common";

export const QuestionTypeSchema = z.enum([
  "Slider",
  "ChoicesWithSingleAnswer",
  "ChoicesWithMultipleAnswers",
  "YesNo",
  "MultipleText",
  "HowLongAgo",
  "Branch",
  "BranchWithRelativeComparison",
]);

const BaseQuestionSchema = z.object({
  id: QuestionIdSchema,
  type: QuestionTypeSchema,
  question: z.string(),
  next: QuestionIdSchemaNullable(["next"]),
});

export const SliderQuestionSchema = BaseQuestionSchema.extend({
  type: z.literal(QuestionTypeSchema.enum.Slider),
  slider: z.tuple([z.string(), z.string()]), // [left, right]
  defaultValue: z.number().int().nonnegative().max(100).optional(),
  defaultValueFromQuestionId: QuestionIdSchema.optional(),
});

export const ChoiceSchema = z.object({
  key: z.string(),
  value: z.string(),
});

export const ChoicesQuestionSchema = BaseQuestionSchema.extend({
  type: z.union([
    z.literal(QuestionTypeSchema.enum.ChoicesWithSingleAnswer),
    z.literal(QuestionTypeSchema.enum.ChoicesWithMultipleAnswers),
  ]),
  choices: z.array(ChoiceSchema).nonempty(),
  specialCasesStartId: z
    .union([
      // TODO: https://github.com/vriad/zod/issues/104
      // Record<choice key, question ID>
      z.record(QuestionIdSchema.nullable()),
      // For when the user click "Prefer not to answer" or next without option.
      z.object({ _pna: QuestionIdSchema.nullable().optional() }),
    ])
    .optional(),
  randomizeChoicesOrder: z.boolean().optional(),
  randomizeExceptForChoiceIds: z.array(z.string()).optional(),
})
  .refine(
    (question) => {
      if (question.specialCasesStartId && question.choices) {
        const choicesKeys = question.choices.map((choice) => choice.key);
        for (const key of Object.keys(question.specialCasesStartId)) {
          if (key !== "_pna" && !choicesKeys.includes(key)) {
            return false;
          }
        }
      }
      return true;
    },
    {
      message:
        'Keys in `specialCasesStartId` must also be in `choices` or be "_pna".',
      path: ["specialCasesStartId"],
    },
  )
  .refine(
    (question) => {
      if (
        question.randomizeExceptForChoiceIds &&
        !question.randomizeChoicesOrder
      ) {
        return false;
      }
      return true;
    },
    {
      message:
        "`randomizeExceptForChoiceIds` should only be set when " +
        "`randomizeChoicesOrder` is set to `true`.",
      path: ["randomizeExceptForChoiceIds"],
    },
  )
  .refine(
    (question) => {
      if (
        question.choices &&
        question.randomizeExceptForChoiceIds &&
        question.randomizeChoicesOrder
      ) {
        const choicesKeys = question.choices.map((choice) => choice.key);
        for (const exceptKey of question.randomizeExceptForChoiceIds) {
          if (!choicesKeys.includes(exceptKey)) {
            return false;
          }
        }
      }
      return true;
    },
    {
      message:
        "Keys in `randomizeExceptForChoiceIds` should also be present in " +
        "`choices`.",
      path: ["randomizeExceptForChoiceIds"],
    },
  );

export const ChoicesWithSingleAnswerQuestionSchema = ChoicesQuestionSchema.extend(
  {
    type: z.literal(QuestionTypeSchema.enum.ChoicesWithSingleAnswer),
  },
);

export const ChoicesWithMultipleAnswersQuestionSchema = ChoicesQuestionSchema.extend(
  {
    type: z.literal(QuestionTypeSchema.enum.ChoicesWithMultipleAnswers),
  },
);

export const YesNoQuestionSchema = BaseQuestionSchema.extend({
  type: z.literal(QuestionTypeSchema.enum.YesNo),
  branchStartId: z
    .object({
      yes: QuestionIdSchema.nullable().optional(),
      no: QuestionIdSchema.nullable().optional(),
    })
    .optional(),
  // Currently only `yes` is supported and also can only followup after 3 days
  // and 7 days with the same stream.
  addFollowupStream: z
    .object({
      yes: StreamNameSchema.optional(),
      // TODO: no: StreamNameSchema.optional(),
    })
    .optional(),
});

export const MultipleTextQuestionSchema = BaseQuestionSchema.extend({
  // `id` will store the number of text fields answered.
  type: z.literal(QuestionTypeSchema.enum.MultipleText),
  indexName: z.string().nonempty(),
  variableName: z.string().nonempty(),
  placeholder: z.string().optional(),
  choices: z.union([z.literal("NAMES"), z.array(ChoiceSchema)]).optional(),
  forceChoice: z.boolean().optional(),
  max: z.number().int().positive(),
  // The max number of text field will be `max` minus the number of text the
  // participant entered in `maxMinus` question.
  maxMinus: QuestionIdSchema.optional(),
  repeatedItemStartId: QuestionIdSchema.optional(),
  // This is used when the user does not enter any name or select prefer not to
  // answer. Note that this has to exists somewhere else. If it is `null`, we
  // will go to `next` directly.
  // TODO: MAKE THIS null = stop HERE
  fallbackItemStartId: QuestionIdSchema.nullable().optional(),
});

export const HowLongAgoQuestionSchema = BaseQuestionSchema.extend({
  type: z.literal(QuestionTypeSchema.enum.HowLongAgo),
});

export const BranchQuestionSchema = BaseQuestionSchema.extend({
  // This is not actually a question (it will not be displayed to the user)
  type: z.literal(QuestionTypeSchema.enum.Branch),
  condition: z.object({
    questionId: QuestionIdSchema,
    questionType: z.union([
      z.literal(QuestionTypeSchema.enum.MultipleText),
      z.literal(QuestionTypeSchema.enum.ChoicesWithSingleAnswer),
    ]),
    compare: z.literal("equal"),
    target: z.union([z.number(), z.string()]),
  }),
  branchStartId: z.object({
    // TODO: ALSO ALLOW NULL = STOP HERE
    true: QuestionIdSchema.nullable().optional(),
    false: QuestionIdSchema.nullable().optional(),
  }),
});

export const BranchWithRelativeComparisonQuestionSchema = BaseQuestionSchema.extend(
  {
    // This is not actually a question (it will not be displayed to the user)
    type: z.literal(QuestionTypeSchema.enum.BranchWithRelativeComparison),
    // TODO: ALSO ALLOW NULL = STOP HERE
    branchStartId: z.record(QuestionIdSchema.nullable()),
  },
);

export const QuestionSchema = z
  .union([
    SliderQuestionSchema,
    ChoicesWithSingleAnswerQuestionSchema,
    ChoicesWithMultipleAnswersQuestionSchema,
    YesNoQuestionSchema,
    MultipleTextQuestionSchema,
    HowLongAgoQuestionSchema,
    BranchQuestionSchema,
    BranchWithRelativeComparisonQuestionSchema,
  ])
  .refine((question) => {
    /**
     * This is to tell user in more details what is going wrong.
     *
     * `...Schema.parse` should throw error if any, replacing the default
     * error.
     *
     * The `default:` clause should handle the case where the type is invalid.
     */
    switch (question.type) {
      case QuestionTypeSchema.enum.Slider:
        SliderQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.ChoicesWithSingleAnswer:
        ChoicesWithSingleAnswerQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.ChoicesWithMultipleAnswers:
        ChoicesWithMultipleAnswersQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.YesNo:
        YesNoQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.MultipleText:
        MultipleTextQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.HowLongAgo:
        HowLongAgoQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.Branch:
        BranchQuestionSchema.parse(question);
        break;

      case QuestionTypeSchema.enum.BranchWithRelativeComparison:
        BranchWithRelativeComparisonQuestionSchema.parse(question);
        break;

      default:
        BaseQuestionSchema.parse(question);
        break;
    }
    return true;
  });

export const QuestionsListSchema = z.record(QuestionSchema).refine(
  (questions) => {
    for (const questionId in questions) {
      if (questionId !== questions[questionId].id) {
        return false;
      }
    }
    return true;
  },
  {
    message:
      "The key in for the question in questions list should be same as " +
      "its question ID.",
  },
);
