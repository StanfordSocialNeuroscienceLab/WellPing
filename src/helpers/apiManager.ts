import Constants from "expo-constants";
import * as Crypto from "expo-crypto";
import * as Device from "expo-device";
import { Platform } from "react-native";

import { AnswerEntity } from "../entities/AnswerEntity";
import { PingEntity } from "../entities/PingEntity";
import { getPingsAsync, getAnswersAsync } from "./asyncStorage";
import { getSurveyFileAsync } from "./configFiles";
import { getUserAsync, User, storeUserAsync } from "./user";

export async function getServerUrlAsync(): Promise<string> {
  return (await getSurveyFileAsync()).studyInfo.serverURL;
}

type UploadData = {
  patientId: string;
  pings: PingEntity[];
  answers: AnswerEntity[];
};

// If success, return `null`. Else return error message.
export async function registerUserAsync(user: User): Promise<string | null> {
  if (!user.patientId || !user.password) {
    return "You should enter both User ID and password.";
  }

  const request: { [key: string]: any } = {
    new_password: user.password, // Do not reset password
    phone_number: "1234567890",
    device_os: Device.osName,
    os_version: Device.osVersion,
    product: Device.modelName,
    brand: Device.brand,
    manufacturer: Device.manufacturer,
    model: Device.modelId,
    beiwe_version: "1.6.0",
  };

  try {
    let endpoint = "/register_user";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    await makePostRequestAsync(endpoint, request, {}, user);
  } catch (e) {
    return `Request error: ${e}.`;
  }

  await storeUserAsync(user);
  return null;
}

export async function getAllDataAsync(): Promise<UploadData> {
  const user = await getUserAsync();
  const pings = await getPingsAsync();
  const answers = await getAnswersAsync();

  const data = {
    patientId: user!.patientId,
    pings,
    answers,
  };
  return data;
}

export async function uploadDataAsync() {
  const user = (await getUserAsync())!;
  const data = await getAllDataAsync();

  try {
    let endpoint = "/ssnl_upload";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    const response = await makePostRequestAsync(endpoint, {}, data, user);
    //console.warn(`YAYA ${JSON.stringify(response)}`);
    return { status: "success", response };
  } catch (e) {
    return { status: "error", error: `Request error: ${e}.` };
  }
}

function base64ToBase64URL(input: string): string {
  return input.replace(/\+/g, "-").replace(/\//g, "_");
}

export async function getRequestURLAsync(
  endpoint: string,
  request: { [key: string]: any } = {},
  forUser?: User,
): Promise<string> {
  let user: User | null = forUser || null;
  if (!user) {
    user = await getUserAsync();

    if (user == null) {
      throw new Error("user == null in getRequestURLAsync");
    }
  }

  const passwordHash = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    user.password,
    {
      encoding: Crypto.CryptoEncoding.BASE64,
    },
  );
  request["password"] = base64ToBase64URL(passwordHash);
  request["device_id"] = `${user.patientId}-${Constants.installationId}`;
  request["patient_id"] = user.patientId;

  //console.warn(`request is ${JSON.stringify(request)}`);

  const serverUrl = await getServerUrlAsync();
  const query = Object.keys(request)
    .map((k) => encodeURIComponent(k) + "=" + encodeURIComponent(request[k]))
    .join("&");
  const url = `${serverUrl}${endpoint}?${query}`;
  return url;
}

export async function makePostRequestAsync(
  endpoint: string,
  request: { [key: string]: any },
  body?: { [key: string]: any },
  user?: User,
): Promise<any> {
  const headers = {
    "Beiwe-Api-Version": "2",
    Accept: "application/vnd.beiwe.api.v2, application/json",
  };

  const url = await getRequestURLAsync(endpoint, request, user);

  //console.warn("url is " + url);

  const response = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });

  if (response.status < 200 || response.status >= 400) {
    console.warn(`ERRORREQUEST: ${JSON.stringify(response)}`);

    if (response.status === 401 || response.status === 403) {
      throw new Error(
        `Verification failed.\n\nRequest: ${JSON.stringify(
          request,
        )}.\n\nResponse: ${JSON.stringify(response)}.`,
      );
    }

    throw new Error(
      `Request not successful.\n\nRequest: ${JSON.stringify(
        request,
      )}.\n\nResponse: ${JSON.stringify(response)}.`,
    );
  }

  //console.warn(response);

  return response.json();
}
