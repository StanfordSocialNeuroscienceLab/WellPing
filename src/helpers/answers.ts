import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
