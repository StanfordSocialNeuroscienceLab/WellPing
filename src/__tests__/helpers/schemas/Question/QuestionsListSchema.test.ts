import { QuestionType } from "../../../../helpers/helpers";
import { QuestionsListSchema } from "../../../../helpers/schemas/Question";

describe("QuestionsListSchema", () => {
  test("should not accept inconsistent key and id", () => {
    expect(() => {
      QuestionsListSchema.parse({
        Feel_Ideal: {
          id: "Feel_Ideal",
          type: QuestionType.HowLongAgo,
          question: "how long ago question",
          next: "Next_Question",
        },
      });
    }).not.toThrowError();

    expect(() => {
      QuestionsListSchema.parse({
        NotTheSameKey: {
          id: "Feel_Ideal",
          type: QuestionType.HowLongAgo,
          question: "how long ago question",
          next: "Next_Question",
        },
      });
    }).toThrowErrorMatchingSnapshot("single");

    expect(() => {
      QuestionsListSchema.parse({
        NotTheSameKey: {
          id: "Feel_Ideal",
          type: QuestionType.HowLongAgo,
          question: "how long ago question",
          next: "Feel_Current",
        },
        Feel_Current: {
          id: "Feel_Current",
          type: QuestionType.HowLongAgo,
          question: "how long ago question",
          next: "How_Are",
        },
        How_Are: {
          id: "NotId",
          type: QuestionType.HowLongAgo,
          question: "how long ago question",
          next: null,
        },
      });
    }).toThrowErrorMatchingSnapshot("multiple");
  });
});
