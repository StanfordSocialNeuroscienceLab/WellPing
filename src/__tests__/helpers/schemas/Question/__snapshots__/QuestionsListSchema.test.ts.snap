// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionsListSchema should not accept inconsistent key and id: multiple 1`] = `
"1 validation issue(s)

  Issue #0: custom_error at 
  The key for the question in questions list should be same as its question ID.
"
`;

exports[`QuestionsListSchema should not accept inconsistent key and id: single 1`] = `
"1 validation issue(s)

  Issue #0: custom_error at 
  The key for the question in questions list should be same as its question ID.
"
`;
